### PR TITLE
Feature/rh bottom crt 3

### PR DESCRIPTION
**Pull request by ron003**
### Description

Put description of request here. Include _Fixes #XXX_ to link to an issue.

_If the full description is in a PR from a different repo, simply link that here, and delete the rest._

### Related Repository Branches

List related branches from other repos here (links to PRs even better!), if not linked above.

### Testing details
_(Note: edit as appropriate.)_
- *Where it was tested*: SBN-FD
- *Run number associated with test*:  9912
- Other information:

### CODE REVIEW

**Review by aaduszki — COMMENTED** [review]
1. To my understanding the fragment generator manipulates numbers using 4 different time units: ns, s, 16 ns ticks, and numbers of rollovers of the sync signal. It makes it very difficult to follow and debug, especially the critical algorithm to assign the timestamps.

I suggest the following changes:
    - print units in all TRACE messages
    - add units to all variable names, rather than comments, e.g., newtimediff_ticks, rather than newtimediff
    - I would actually suggest converting all numbers (maybe except of the number of rollovers) to nanoseconds, to avoid the ambiguity entirely 

2. Convert all cout and prinft messages to TRACE messages, in particular those indicating failures, to ensure they can be found in standard artdaq logs
  
3. Don't commit .swp temporary files. I would also consider putting the pdf in docdb, and referring to it with a docdb number

File: sbndaq-artdaq/Config/fhicl/BottomCRTDriver.fcl [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  # Exactly one artdaq process needs to start up the backend, but then
+  # all four can read the results
+  startbackend: true
+
+  # valid number 39
+  usbnumber: 39
+
+  # Options are "crt_upstream", "crt_downstream", "online_normal_running"
+  #sqltable: "board_6_debug"
+  configfile: "/home/nfs/icarus/DAQ_DevAreas/DAQ_12Dec2022_rhowell/srcs/sbndaq_artdaq/sbndaq-artdaq/Generators/ICARUS/BottomInterface/Backend_DAQ/DAQ_CPP_v1/crt_daq.fcl"
```
Comment (aaduszki) [inline sbndaq-artdaq/Config/fhicl/BottomCRTDriver.fcl:30]: Use of absolute path is likely to cause issues with maintaining the code in the future.

**Comment by wesketchum** [comment]
I've tried to give this a general look-over, and beyond Antoni's comments both here and in the email he sent around I don't have too much to say. A few general things:

* It's unclear to me the need for a separate fcl file for the channel/gain configurations, which has its own file reader for the fcl format. Is it possible to have an option to use the fcl libraries for reading a parameter set, and have an option to pass information in as a parameter set instead?

As it is, I'm a little worried about how we store this additional fcl file for configuration and update it if needed (I suppose it can be included in an include path ... but it clearly would be better to store in the configuration with everything else)

* It's not clear to me looking at the code what will happen if the bottomCRTreadout fails to start or if there is some error in the interaction via the interface program. If it won't cause a clear error / throw a clear exception, then I'd propose that we try to add something like that in...maybe, as an example, what happens if the readout driver program crashes in the middle of the run? How would we diagnose this/make it clear to the shifter? If there's not a good mechanism for that, perhaps we can add checks on the getNext_ loop (or in a separate monitoring thread) to monitor to make sure that process is there and issue an error message if not?

* There are a number of places where we may try to cleanup / improve the code. Antoni's idea to try to bring in the use of the C libraries so that we don't have to run a separate process is a good one (and one of the intentions of doing the work to update this to C, so thank you for that!). Perhaps this should not get in the way of what we want to do for initial integration testing, but could be done on the time-scale of this summer. It could perhaps be coupled with changes/development for getting a similar system up and running in ProtoDUNE-II.

* As we all know, before merging, we should make sure this looks like it runs with the rest of the system, produces data fragments with a low enough latency, that we have tuned all the timeouts that are needed, and hopefully have some validation that the data is properly synchronized prior to merge. The latter may require more data, so as long as there is stable running of the system, then I think it may be ok. Likely for initial running, we should make sure CRT bottom DAQ experts are available/on-hand to help with debugging issues if problems from the shifter show up (that is, may want a special on-call expert there until things are fully stable/understood).

**Review by aaduszki — APPROVED** [review]
While this is work in progress, but this version has been tested and it doesn't interfere with regular DAQ operation, we can merge it for now, while the development continues.

**Pull request merged**